File: rust/2015/day09a/src/main.rs
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;

use itertools::Itertools;
use regex::Regex;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn parse_file(filename: &str) -> io::Result<Vec<(String, String, u32)>> {
    let line_re = Regex::new(r"([A-Za-z]+) to ([A-Za-z]+) = (\d+)").unwrap();
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    let mut input_data: Vec<(String, String, u32)> = vec![];
    for line_result in reader.lines() {
        let line = line_result?;
        if let Some(caps) = line_re.captures(&line) {
            let from = String::from(&caps[1]);
            let to = String::from(&caps[2]);
            let dist = caps[3].parse::<u32>().unwrap();
            input_data.push((from, to, dist));
        }
    }
    Ok(input_data)
}

fn build_map(edges: &[(String, String, u32)]) -> HashMap<(String, String), u32> {
    let mut distances: HashMap<(String, String), u32> = HashMap::new();
    edges.iter().for_each(|(c1, c2, d)| {
        distances.insert((c1.clone(), c2.clone()), *d);
        distances.insert((c2.clone(), c1.clone()), *d);
    });
    distances
}

fn path_len(distances: &HashMap<(String, String), u32>, path: &[&String]) -> u32 {
    path.windows(2)
        .map(|cs| {
            let c1 = cs[0];
            let c2 = cs[1];
            distances.get(&(c1.clone(), c2.clone())).unwrap()
        })
        .sum()
}

fn shortest_route(cities: &[String], distances: &HashMap<(String, String), u32>) -> u32 {
    cities
        .iter()
        .permutations(cities.len())
        .map(|path| path_len(distances, &path))
        .min()
        .unwrap()
}

fn process_file(filename: &str) -> io::Result<u32> {
    let edges = parse_file(filename)?;
    let distances = build_map(&edges);
    let mut tmp_set: HashSet<String> = HashSet::new();
    distances.keys().for_each(|(c1, c2)| {
        tmp_set.insert(c1.clone());
        tmp_set.insert(c2.clone());
    });
    let cities: Vec<String> = tmp_set.into_iter().collect();
    let result = shortest_route(&cities, &distances);
    Ok(result)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}

File: haskell/2015/day09a/src/Main.hs
module Main ( main ) where

import Control.DeepSeq
import Data.Int (Int32)
import Data.List (permutations)
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import System.Clock
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Text.Parsec
import Text.Parsec.String (Parser)

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

type City = String
type Distances = Map.Map (City, City) Int

city :: Parser City
city = many1 letter

line :: Parser (City, City, Int)
line = do
       city1 <- city
       spaces
       _ <- string "to"
       spaces
       city2 <- city
       spaces
       _ <- string "="
       spaces
       distance <- many1 digit
       return (city1, city2, read distance)

file :: Parser [(City, City, Int)]
file = line `endBy1` newline

buildMap :: [(City, City, Int)] -> Distances
buildMap edges = Map.fromList [((a, b), n) | (a, b, n) <- edges ] <>
                 Map.fromList [((b, a), n) | (a, b, n) <- edges ]

dist :: Distances -> City -> City -> Int
dist d a b = d Map.! (a, b)

pathLen :: Distances -> [City] -> Int
pathLen d cs = sum [ dist d a b | (a, b) <- zip cs (tail cs) ]

shortestRoute :: [City] -> Distances -> Int
shortestRoute cities d = minimum [ pathLen d p | p <- permutations cities ]

process :: String -> Int32
process content =
    let edges = case parse file "" content of
                    Left err -> error (show err)
                    Right xs -> xs
        distances = buildMap edges
        cities = Set.toList (Set.fromList [ c | (a, b, _) <- edges, c <- [a, b] ])
    in fromIntegral $ shortestRoute cities distances

showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname
